File: packages/python-sdk/src/webmcp_bridge/drivers/playwright_driver.py
"""
PlaywrightDriver — Python implementation of BridgeDriver using Playwright.

Maps bridge operations to Playwright's sync API:
- find_element → page.locator() with strategy-based selectors
- click → locator.click()
- type_text → locator.fill()
- read_text → locator.text_content()
- screenshot → page.screenshot()

Implementation notes for agents:
- Constructor takes a Playwright Page (sync), Browser, and BrowserContext
- ARIA strategy → page.get_by_role(role, name=name)
- Label strategy → page.get_by_label(text)
- Text strategy → page.get_by_text(text)
- CSS strategy → page.locator(selector)
- JS strategy → page.evaluate(expression)
"""
from __future__ import annotations

import re
from typing import Any


def _escape_regex(s: str) -> str:
    """Escape special regex characters in a string."""
    return re.escape(s)


class PlaywrightDriver:
    """Playwright-based BridgeDriver implementation."""

    def __init__(
        self,
        page: Any,
        browser: Any | None = None,
        context: Any | None = None,
        timeout: int = 30000,
    ) -> None:
        self._page = page
        self._browser = browser
        self._context = context
        self._timeout = timeout
        self._named_pages: dict[str, Any] = {}

    # ─── Navigation ────────────────────────────────────

    def goto(self, url: str) -> None:
        """Navigate to URL."""
        self._page.goto(url, wait_until="domcontentloaded")

    def wait_for_navigation(self, url_pattern: str | None = None) -> None:
        """Wait for navigation to complete."""
        if url_pattern:
            self._page.wait_for_url(url_pattern, timeout=self._timeout)
        else:
            self._page.wait_for_load_state("domcontentloaded", timeout=self._timeout)

    # ─── Element Finding ──────────────────────────────

    def find_element(self, selectors: list[dict[str, Any]]) -> Any:
        """Find element using selector chain with strategy fallback."""
        for strategy in selectors:
            try:
                locator = self._map_strategy_to_locator(strategy)
                if locator is None:
                    continue
                locator.first.wait_for(state="attached", timeout=5000)
                return locator.first
            except Exception:
                continue
        raise RuntimeError("Failed to find element with any strategy")

    def _map_strategy_to_locator(self, strategy: dict[str, Any]) -> Any:
        """Map a selector strategy to a Playwright locator."""
        kind = strategy["strategy"]

        if kind == "css":
            return self._page.locator(strategy["selector"])

        if kind == "aria":
            name = strategy.get("name")
            name_pattern = (
                re.compile(_escape_regex(name), re.IGNORECASE) if name else None
            )
            return self._page.get_by_role(strategy["role"], name=name_pattern)

        if kind == "label":
            text_pattern = re.compile(_escape_regex(strategy["text"]), re.IGNORECASE)
            return self._page.get_by_label(text_pattern)

        if kind == "text":
            if strategy.get("exact"):
                return self._page.get_by_text(strategy["text"], exact=True)
            text_pattern = re.compile(_escape_regex(strategy["text"]), re.IGNORECASE)
            return self._page.get_by_text(text_pattern)

        if kind == "js":
            element = self._page.evaluate(strategy["expression"])
            if element:
                return element
            return None

        msg = f"Unknown selector strategy: {kind}"
        raise ValueError(msg)

    # ─── Interactions ─────────────────────────────────

    def click(self, element: Any) -> None:
        """Click element."""
        element.click(timeout=self._timeout)

    def double_click(self, element: Any) -> None:
        """Double-click element."""
        element.dblclick(timeout=self._timeout)

    def type_text(
        self,
        element: Any,
        text: str,
        clear: bool = False,
        delay: int | None = None,
    ) -> None:
        """Type text into element."""
        if clear:
            element.clear()
        if delay:
            element.press_sequentially(text, delay=delay)
        else:
            element.fill(text)

    def select(self, element: Any, value: str) -> None:
        """Select option from dropdown."""
        element.select_option(value)

    def check(self, element: Any, state: bool) -> None:
        """Check or uncheck a checkbox."""
        if state:
            element.check(timeout=self._timeout)
        else:
            element.uncheck(timeout=self._timeout)

    def clear(self, element: Any) -> None:
        """Clear element value."""
        element.clear()

    def hover(self, element: Any) -> None:
        """Hover over element."""
        element.hover(timeout=self._timeout)

    def drag_drop(self, source: Any, target: Any) -> None:
        """Drag source element to target element."""
        source.drag_to(target)

    def upload_file(self, element: Any, paths: list[str]) -> None:
        """Upload files to input element."""
        element.set_input_files(paths)

    # ─── Reading ──────────────────────────────────────

    def read_text(self, selectors: list[dict[str, Any]]) -> str:
        """Read text from element found by selectors."""
        element = self.find_element(selectors)
        text = element.text_content()
        return (text or "").strip()

    def read_pattern(
        self, selectors: list[dict[str, Any]], regex: str
    ) -> str | None:
        """Read text and match a regex pattern."""
        text = self.read_text(selectors)
        pattern = re.compile(regex)
        match = pattern.search(text)
        if not match:
            return None
        return match.group(1) if match.lastindex and match.lastindex >= 1 else match.group(0)

    # ─── Keyboard ─────────────────────────────────────

    def press_key(self, key: str, modifiers: list[str] | None = None) -> None:
        """Press a keyboard key with optional modifiers."""
        if modifiers:
            combo = "+".join(modifiers) + "+" + key
        else:
            combo = key
        self._page.keyboard.press(combo)

    def press_sequentially(
        self, element: Any, text: str, delay: int | None = None
    ) -> None:
        """Type text character by character."""
        element.press_sequentially(text, delay=delay)

    # ─── Scrolling ────────────────────────────────────

    def scroll(
        self,
        target: Any | str | None = None,
        behavior: str = "instant",
    ) -> None:
        """Scroll page or element into view."""
        if isinstance(target, str) and target == "top":
            self._page.evaluate(f"window.scrollTo({{ top: 0, behavior: '{behavior}' }})")
        elif isinstance(target, str) and target == "bottom":
            self._page.evaluate(
                f"window.scrollTo({{ top: document.body.scrollHeight, behavior: '{behavior}' }})"
            )
        elif target is not None and not isinstance(target, str):
            target.scroll_into_view_if_needed(timeout=self._timeout)
        else:
            self._page.evaluate(f"window.scrollTo({{ top: 0, behavior: '{behavior}' }})")

    # ─── Overlays ─────────────────────────────────────

    def dismiss_overlay(self, strategy: dict[str, Any]) -> bool:
        """Try to dismiss an overlay using the given strategy."""
        try:
            overlay_type = strategy["type"]
            if overlay_type == "press_escape":
                self._page.keyboard.press("Escape")
            elif overlay_type == "click_close":
                selector = strategy.get("selector")
                if selector:
                    self._page.locator(selector).click(timeout=3000)
            elif overlay_type == "click_text":
                texts = strategy.get("text", [])
                for t in texts:
                    try:
                        self._page.get_by_text(t, exact=False).click(timeout=2000)
                        break
                    except Exception:
                        continue
            elif overlay_type == "remove_element":
                selector = strategy.get("selector")
                if selector:
                    self._page.evaluate(
                        f"document.querySelector('{selector}')?.remove()"
                    )

            wait_after = strategy.get("waitAfter")
            if wait_after:
                self._page.wait_for_timeout(wait_after)
            return True
        except Exception:
            return False

    # ─── Events ───────────────────────────────────────

    def dispatch_event(
        self, element: Any, event: str, detail: Any | None = None
    ) -> None:
        """Dispatch a DOM event on an element."""
        event_init = {"detail": detail} if detail else None
        element.dispatch_event(event, event_init)

    # ─── Dialogs ──────────────────────────────────────

    def handle_dialog(
        self, action: str, prompt_text: str | None = None
    ) -> str:
        """Handle a JavaScript dialog (alert/confirm/prompt)."""
        message = ""

        def on_dialog(dialog: Any) -> None:
            nonlocal message
            message = dialog.message
            if action == "accept":
                dialog.accept(prompt_text)
            else:
                dialog.dismiss()

        self._page.once("dialog", on_dialog)
        return message

    # ─── Waiting ──────────────────────────────────────

    def wait_for(self, condition: dict[str, Any]) -> None:
        """Wait for a condition to be met."""
        cond_type = condition["type"]
        value = condition.get("value")
        timeout = condition.get("timeout", self._timeout)

        if cond_type == "selector":
            self._page.locator(value).wait_for(state="visible", timeout=timeout)
        elif cond_type == "url":
            self._page.wait_for_url(value, timeout=timeout)
        elif cond_type == "timeout":
            self._page.wait_for_timeout(value)
        elif cond_type == "network_idle":
            self._page.wait_for_load_state("networkidle", timeout=timeout)
        else:
            msg = f"Unknown wait condition type: {cond_type}"
            raise ValueError(msg)

    # ─── Diagnostics ──────────────────────────────────

    def screenshot(self) -> bytes:
        """Take a full-page screenshot."""
        return self._page.screenshot(full_page=True)  # type: ignore[no-any-return]

    def evaluate(self, js: str) -> Any:
        """Evaluate JavaScript in the page context."""
        return self._page.evaluate(js)

    # ─── Context ──────────────────────────────────────

    def get_page_context(self) -> dict[str, Any]:
        """Get current page context information."""
        return {
            "url": self._page.url,
            "title": self._page.title(),
            "readyState": self._page.evaluate("document.readyState"),
        }

    # ─── Multi-tab ────────────────────────────────────

    def get_named_page(self, name: str) -> Any:
        """Get a previously created named page."""
        if name not in self._named_pages:
            msg = f"Page not found: {name}"
            raise RuntimeError(msg)
        page = self._named_pages[name]
        page.bring_to_front()
        return page

    def create_page(self, name: str) -> Any:
        """Create a new named page/tab."""
        if self._context is None:
            msg = "No browser context available"
            raise RuntimeError(msg)
        page = self._context.new_page()
        self._named_pages[name] = page
        return page

    # ─── Legacy compatibility ─────────────────────────

    def wait_for_selector(self, selector: str, timeout: int = 10000) -> None:
        """Wait for a CSS selector to be visible (legacy method)."""
        self.wait_for({"type": "selector", "value": selector, "timeout": timeout})

File: packages/playwright/src/playwright-driver.ts
/**
 * PlaywrightDriver — BridgeDriver implementation using Playwright.
 *
 * Maps each BridgeDriver method to Playwright's Page API:
 * - findElement → page.locator() with strategy-specific selectors
 * - click → locator.click()
 * - type → locator.fill() or locator.pressSequentially()
 * - readText → locator.textContent()
 * - screenshot → page.screenshot()
 * - etc.
 *
 * Implementation notes for agents:
 * - Constructor takes a Playwright Page instance
 * - ElementHandle wraps Playwright Locator (not Playwright ElementHandle, which is deprecated)
 * - For ARIA strategy: use page.getByRole(role, { name })
 * - For label strategy: use page.getByLabel(text)
 * - For text strategy: use page.getByText(text, { exact })
 * - For CSS strategy: use page.locator(selector)
 * - For JS strategy: use page.evaluate(expression)
 * - Multi-tab: use browser.contexts()[0].pages() for named pages
 */
import type {
  BridgeDriver,
  ElementHandle,
  PageHandle,
  PageContext,
  SelectorChain,
  SelectorStrategy,
  TypeOpts,
  ScrollOpts,
  DismissStrategy,
  WaitCondition,
  EventOpts,
} from '@webmcp-bridge/core';

import type { Page, Browser, BrowserContext, Locator } from 'playwright';

// Internal interface to carry the Playwright Locator inside an ElementHandle
interface LocatorElementHandle extends ElementHandle {
  readonly __locator: Locator;
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function wrapLocator(locator: Locator): ElementHandle {
  return { _brand: 'ElementHandle', __locator: locator } as unknown as ElementHandle;
}

function unwrapLocator(element: ElementHandle): Locator {
  return (element as unknown as LocatorElementHandle).__locator;
}

export class PlaywrightDriver implements BridgeDriver {
  private page: Page;
  private browser: Browser;
  private context: BrowserContext;
  private timeout: number;
  private namedPages: Map<string, Page> = new Map();

  constructor(page: Page, browser: Browser, context: BrowserContext, timeout: number = 30000) {
    this.page = page;
    this.browser = browser;
    this.context = context;
    this.timeout = timeout;
  }

  private mapStrategyToLocator(strategy: SelectorStrategy): Locator | null {
    switch (strategy.strategy) {
      case 'aria':
        return this.page.getByRole(strategy.role as Parameters<Page['getByRole']>[0], {
          name: strategy.name ? new RegExp(escapeRegex(strategy.name), 'i') : undefined,
        });
      case 'label':
        return this.page.getByLabel(new RegExp(escapeRegex(strategy.text), 'i'));
      case 'text':
        return strategy.exact
          ? this.page.getByText(strategy.text, { exact: true })
          : this.page.getByText(new RegExp(escapeRegex(strategy.text), 'i'));
      case 'css':
        return this.page.locator(strategy.selector);
      case 'js':
        return null; // handled specially
      default:
        throw new Error(`Unknown selector strategy: ${(strategy as { strategy: string }).strategy}`);
    }
  }

  async goto(url: string): Promise<void> {
    await this.page.goto(url, { waitUntil: 'domcontentloaded' });
  }

  async waitForNavigation(urlPattern?: string | RegExp): Promise<void> {
    if (urlPattern) {
      await this.page.waitForURL(urlPattern, { timeout: this.timeout });
    } else {
      await this.page.waitForLoadState('domcontentloaded', { timeout: this.timeout });
    }
  }

  async findElement(selectors: SelectorChain): Promise<ElementHandle> {
    for (const strategy of selectors) {
      try {
        if (strategy.strategy === 'js') {
          const handle = await this.page.evaluateHandle(strategy.expression);
          const element = handle.asElement();
          if (element) {
            // Convert ElementHandle to a locator-like wrapper
            // We use page.locator with a JS expression
            const locator = this.page.locator(`js=${strategy.expression}`).first();
            try {
              await locator.waitFor({ state: 'attached', timeout: 5000 });
              return wrapLocator(locator);
            } catch {
              // If locator approach fails, fall through
            }
          }
          continue;
        }

        const locator = this.mapStrategyToLocator(strategy);
        if (!locator) continue;

        await locator.first().waitFor({ state: 'attached', timeout: 5000 });
        return wrapLocator(locator.first());
      } catch {
        continue;
      }
    }
    throw new Error('Failed to find element with any strategy');
  }

  async click(element: ElementHandle): Promise<void> {
    await unwrapLocator(element).click({ timeout: this.timeout });
  }

  async doubleClick(element: ElementHandle): Promise<void> {
    await unwrapLocator(element).dblclick({ timeout: this.timeout });
  }

  async type(element: ElementHandle, text: string, opts?: TypeOpts): Promise<void> {
    const locator = unwrapLocator(element);
    if (opts?.clear) {
      await locator.clear();
    }
    if (opts?.delay) {
      await locator.pressSequentially(text, { delay: opts.delay });
    } else {
      await locator.fill(text);
    }
  }

  async select(element: ElementHandle, value: string): Promise<void> {
    await unwrapLocator(element).selectOption(value);
  }

  async check(element: ElementHandle, state: boolean): Promise<void> {
    const locator = unwrapLocator(element);
    if (state) {
      await locator.check({ timeout: this.timeout });
    } else {
      await locator.uncheck({ timeout: this.timeout });
    }
  }

  async clear(element: ElementHandle): Promise<void> {
    await unwrapLocator(element).clear();
  }

  async hover(element: ElementHandle): Promise<void> {
    await unwrapLocator(element).hover({ timeout: this.timeout });
  }

  async dragDrop(source: ElementHandle, target: ElementHandle): Promise<void> {
    await unwrapLocator(source).dragTo(unwrapLocator(target));
  }

  async uploadFile(input: ElementHandle, paths: string[]): Promise<void> {
    await unwrapLocator(input).setInputFiles(paths);
  }

  async readText(selectors: SelectorChain): Promise<string> {
    const element = await this.findElement(selectors);
    const locator = unwrapLocator(element);
    const text = await locator.textContent({ timeout: this.timeout });
    return text?.trim() || '';
  }

  async readPattern(selectors: SelectorChain, regex: string): Promise<string | null> {
    const text = await this.readText(selectors);
    const pattern = new RegExp(regex);
    const match = pattern.exec(text);
    if (!match) return null;
    return match[1] ?? match[0];
  }

  async pressKey(key: string, modifiers?: string[]): Promise<void> {
    const combo = modifiers?.length ? `${modifiers.join('+')}+${key}` : key;
    await this.page.keyboard.press(combo);
  }

  async pressSequentially(element: ElementHandle, text: string, delay?: number): Promise<void> {
    await unwrapLocator(element).pressSequentially(text, delay ? { delay } : undefined);
  }

  async scroll(target?: ElementHandle | 'top' | 'bottom', opts?: ScrollOpts): Promise<void> {
    const behavior = opts?.behavior || 'instant';
    if (target === 'top') {
      await this.page.evaluate(`window.scrollTo({ top: 0, behavior: '${behavior}' })`);
    } else if (target === 'bottom') {
      await this.page.evaluate(`window.scrollTo({ top: document.body.scrollHeight, behavior: '${behavior}' })`);
    } else if (target) {
      await unwrapLocator(target as ElementHandle).scrollIntoViewIfNeeded({ timeout: this.timeout });
    } else {
      await this.page.evaluate(`window.scrollTo({ top: 0, behavior: '${behavior}' })`);
    }
  }

  async dismissOverlay(strategy: DismissStrategy): Promise<boolean> {
    try {
      switch (strategy.type) {
        case 'press_escape':
          await this.page.keyboard.press('Escape');
          break;
        case 'click_close':
          if (strategy.selector) {
            await this.page.locator(strategy.selector).click({ timeout: 3000 });
          }
          break;
        case 'click_text':
          if (strategy.text?.length) {
            for (const t of strategy.text) {
              try {
                await this.page.getByText(t, { exact: false }).click({ timeout: 2000 });
                break;
              } catch { continue; }
            }
          }
          break;
        case 'remove_element':
          if (strategy.selector) {
            await this.page.evaluate(`document.querySelector('${strategy.selector}')?.remove()`);
          }
          break;
      }
      if (strategy.waitAfter) {
        await this.page.waitForTimeout(strategy.waitAfter);
      }
      return true;
    } catch {
      return false;
    }
  }

  async dispatchEvent(element: ElementHandle, event: string, opts?: EventOpts): Promise<void> {
    const locator = unwrapLocator(element);
    await locator.dispatchEvent(event, opts?.detail ? { detail: opts.detail } : undefined);
  }

  async switchFrame(target: string | ElementHandle | 'parent'): Promise<void> {
    if (target === 'parent') {
      // Switch back to main frame
      this.page = this.page.mainFrame().page();
    } else if (typeof target === 'string') {
      const frame = this.page.frameLocator(target);
      // Store frame reference for subsequent operations
      // Playwright uses frameLocator chaining rather than frame switching
      // This is a simplification — real usage would need frame-aware locators
      void frame;
    } else {
      const locator = unwrapLocator(target);
      const frame = this.page.frameLocator(`iframe`).first();
      void frame;
      void locator;
    }
  }

  async handleDialog(action: 'accept' | 'dismiss', promptText?: string): Promise<string> {
    const dialogPromise = this.page.waitForEvent('dialog', { timeout: this.timeout });
    const dialog = await dialogPromise;
    const message = dialog.message();
    if (action === 'accept') {
      await dialog.accept(promptText);
    } else {
      await dialog.dismiss();
    }
    return message;
  }

  async waitFor(condition: WaitCondition): Promise<void> {
    switch (condition.type) {
      case 'selector':
        await this.page.locator(condition.value as string).waitFor({
          state: 'visible',
          timeout: condition.timeout || this.timeout,
        });
        break;
      case 'url':
        await this.page.waitForURL(condition.value as string | RegExp, {
          timeout: condition.timeout || this.timeout,
        });
        break;
      case 'timeout':
        await this.page.waitForTimeout(condition.value as number);
        break;
      case 'network_idle':
        await this.page.waitForLoadState('networkidle', {
          timeout: condition.timeout || this.timeout,
        });
        break;
      default:
        throw new Error(`Unknown wait condition type: ${(condition as { type: string }).type}`);
    }
  }

  async screenshot(): Promise<Buffer> {
    return await this.page.screenshot({ fullPage: true }) as Buffer;
  }

  async evaluate(js: string): Promise<unknown> {
    return await this.page.evaluate(js);
  }

  async getPageContext(): Promise<PageContext> {
    return {
      url: this.page.url(),
      title: await this.page.title(),
      readyState: await this.page.evaluate('document.readyState') as PageContext['readyState'],
    };
  }

  async getNamedPage(name: string): Promise<PageHandle> {
    const page = this.namedPages.get(name);
    if (!page) throw new Error(`Page not found: ${name}`);
    await page.bringToFront();
    return { name, _brand: 'PageHandle' } as PageHandle;
  }

  async createPage(name: string): Promise<PageHandle> {
    const newPage = await this.context.newPage();
    this.namedPages.set(name, newPage);
    return { name, _brand: 'PageHandle' } as PageHandle;
  }
}
